### ISSUE TITLE
Support columns having the same alias

### ISSUE THREAD

**Issue by berkaysynnada** [apache/datafusion#6543]
### Describe the bug

When we give the same aliases for multiple columns (`SELECT ts as c1, inc_col as c1 FROM annotated_data_infinite`), builder gives such an error: 
`Plan("Projections require unique expression names but the expression \"annotated_data_infinite.ts AS c1\" at position 0 and \"annotated_data_infinite.inc_col AS c1\" at position 1 have the same name. Consider aliasing (\"AS\") one of them.")`

### To Reproduce

```
        ctx.sql(
            "CREATE EXTERNAL TABLE annotated_data_infinite (
              ts INTEGER,
              inc_col INTEGER,
              desc_col INTEGER,
            )
            STORED AS CSV
            WITH HEADER ROW
            WITH ORDER (ts ASC)
            LOCATION '/Users/berkaysahin/Desktop/arrow-datafusion/datafusion/core/tests/data/window_1.csv'",
        )
        .await?;
        let sql = "SELECT ts as c1, inc_col as c1 FROM annotated_data_infinite";
        let dataframe = ctx.sql(sql).await.expect(&msg);
```

### Expected behavior

Postgre can handle it and gives result with two columns having the same name. I don't know this is an intentional behaviour in Datafusion or a bug, but I would like to open an issue.

**Comment by comphead** [comment]
Thanks @berkaysynnada for raising. This is quite old problem,DF has unique column name check in the planner.  We planned to move this check one level upper, so the query will fail if outer query references the inner query containing duplicated aliases. Fir you scenario is it real world one?

**Comment by berkaysynnada** [comment]
> Thanks @berkaysynnada for raising. This is quite old problem,DF has unique column name check in the planner. We planned to move this check one level upper, so the query will fail if outer query references the inner query containing duplicated aliases. Fir you scenario is it real world one?

Not actually, it was a hypothetical trial. Your plan makes sense and thanks for letting me know about it.

**Comment by alamb** [comment]
I also filed https://github.com/apache/arrow-datafusion/issues/6758 to think about the problem with large column names.

**Comment by alamb** [comment]
I wonder if some potential solution for this issue would be to automatically add a string to make the columns unique in the arrow schema?

**Comment by comphead** [comment]
that is good idea btw, currently we got bunch of issues
- Arrow schema uniqueness violation
- DFSchema uniqueness violation
this solution can potentially address both issues

**Comment by jackwener** [comment]
This is a legacy issue.

Generally, we won't raise an error for having columns with the same name unless an outer subquery references that column name.

In terms of this issue itself, we should fix it in the planner.

**Comment by alamb** [comment]
🤯  I didn't realize this worked

```sql
(arrow_dev) alamb@MacBook-Pro-8:~/Software/influxdb_iox2$ datafusion-cli
DataFusion CLI v26.0.0
❯ create table foo (x int) as values (1), (2), (3);
0 rows in set. Query took 0.003 seconds.
❯ select x as "my_col", x as "my col" from foo;
+--------+--------+
| my_col | my col |
+--------+--------+
| 1      | 1      |
| 2      | 2      |
| 3      | 3      |
+--------+--------+
3 rows in set. Query took 0.005 seconds.
❯ select x as "my_col", x+1 as "my col" from foo;
+--------+--------+
| my_col | my col |
+--------+--------+
| 1      | 2      |
| 2      | 3      |
| 3      | 4      |
```

However, using `c1` as the alias for some reason fails:

```sql
❯ select x as c1, x as c1 from foo;
Error during planning: Projections require unique expression names but the expression "foo.x AS c1" at position 0 and "foo.x AS c1" at position 1 have the same name. Consider aliasing ("AS") one of them.+--------+--------+
```

**Comment by comphead** [comment]
> 🤯 I didn't realize this worked
> 
> ```sql
> (arrow_dev) alamb@MacBook-Pro-8:~/Software/influxdb_iox2$ datafusion-cli
> DataFusion CLI v26.0.0
> ❯ create table foo (x int) as values (1), (2), (3);
> 0 rows in set. Query took 0.003 seconds.
> ❯ select x as "my_col", x as "my col" from foo;
> +--------+--------+
> | my_col | my col |
> +--------+--------+
> | 1      | 1      |
> | 2      | 2      |
> | 3      | 3      |
> +--------+--------+
> 3 rows in set. Query took 0.005 seconds.
> ❯ select x as "my_col", x+1 as "my col" from foo;
> +--------+--------+
> | my_col | my col |
> +--------+--------+
> | 1      | 2      |
> | 2      | 3      |
> | 3      | 4      |
> ```
> 
> However, using `c1` as the alias for some reason fails:
> 
> ```sql
> ❯ select x as c1, x as c1 from foo;
> Error during planning: Projections require unique expression names but the expression "foo.x AS c1" at position 0 and "foo.x AS c1" at position 1 have the same name. Consider aliasing ("AS") one of them.+--------+--------+
> ```

its failed as DF has projection uniqueness column name check.
first test didn't fail because it has different aliases

**Comment by alamb** [comment]
> first test didn't fail because it has different aliases

Oh man, 🤦  -- I missed the `_` and ` `

**Comment by tv42** [comment]
An extra motivation to get this right is that sqlite's sqllogictest suite has a *lot* of these. You'll get better test coverage if you support this.

**Comment by alamb** [comment]
I wonder if someone has time to file a ticket with the idea to re(use) sqlite's sqllogictest suite?

**Comment by Jefffrey** [comment]
Another case is when selecting same value literals

```sql
DataFusion CLI v37.0.0
❯ select 1, 1;
Error during planning: Projections require unique expression names but the expression "Int64(1)" at position 0 and "Int64(1)" at position 1 have the same name. Consider aliasing ("AS") one of them.
❯
```

**Comment by comphead** [comment]
Yes,I  was thinking the other day to allow query like that and check uniqueness from outer queries only

**Comment by tv42** [comment]
@alamb I don't think you can directly run SQLite's test suite against just datafusion, there's a lot of `CREATE INDEX` etc going on.

I have an early-state OLTP database project using datafusion and *that* survives a decent fraction of sqllogictest. This issue is one of the remaining big limitations, along with `AVG(DISTINCT)` https://github.com/apache/arrow-datafusion/issues/2408 -- I've been filing bugs on test cases that failed due to datafusion, and they've largely been fixed.

**Comment by findepi** [comment]
This would be great item in the broader scope of https://github.com/apache/datafusion/issues/12723, which intents to make DataFusion Logical Plans be state of the art.

**Comment by alamb** [comment]
> what can we do about this ?
> can i update the slt test for this test ?

Whatever is rewriting the function call probably should add an alias 🤔

**Comment by laisee** [comment]
Note extra use case that fails (silently) when joining 2 tables from different data connectors with common column name. Adding here as the error may not be detected i.e. no error code is given and 'wrong' data appears in results 
e.g. `select 
    movies.id as "id", 
    movies.name, 
    coin.name
from 
    movies, coin
where
    movies.id = coin.id`  

This query runs Ok but data selected into result set columns 2 & 3 is from table 'coin'. Reverse the order of "movies.name" and "coin.name" and you will see movies 'name' data in both result set columns.

Solution is to add an alias for columns to make a unique result set by column name.

**Comment by osipovartem** [comment]
Here is the same issue

```sql
CREATE TABLE test (col1 DATE,  col2 DATE);
INSERT INTO test
  SELECT TO_DATE('2013-05-08T23:39:20.123'), TO_DATE('2013-05-08T23:39:20.123')
```
```rust
Projections require unique expression names but the expression "to_date(Utf8("2013-05-08T23:39:20.123"))" at position 0 and "to_date(Utf8("2013-05-08T23:39:20.123"))" at position 1 have the same name. Consider aliasing ("AS") one of them.
```
would it be correct to use statement visitors here to add unique aliases?
Seems we cannot use any analyzer/optimizer rules because the plan builder is failing with a unique expression name error.

```rust
#[derive(Debug, Default)]
pub struct AddAliasesToSelectExpressions {}

impl VisitorMut for AddAliasesToSelectExpressions {
    type Break = ();

    fn pre_visit_query(&mut self, query: &mut Query) -> ControlFlow<Self::Break> {
        if let SetExpr::Select(select) = &mut *query.body {
            let mut seen = HashSet::new();
            let mut counter = 0;

            for item in &mut select.projection {
                match item {
                    SelectItem::UnnamedExpr(expr) => {
                        let expr_str = expr.to_string();

                        if !seen.insert(expr_str.clone()) {
                            let alias = format!("expr_{counter}");
                            *item = SelectItem::ExprWithAlias {
                                expr: expr.clone(),
                                alias: Ident::new(alias),
                            };
                            counter += 1;
                        }
                    }
                    SelectItem::ExprWithAlias { alias, .. } => {
                        seen.insert(alias.value.clone());
                    }
                    _ => {}
                }
            }
        }
        ControlFlow::Continue(())
    }
}

pub fn visit(stmt: &mut Statement) {
    let _ = stmt.visit(&mut AddAliasesToSelectExpressions {});
}
```
or using visit_statements_mut

**Comment by alamb** [comment]
> would it be correct to use statement visitors here to add unique aliases?


I think that is actually a pretty neat idea -- specifically add the aliases in the SQL planner 👍 

I would be happy to review such a PR

**Comment by osipovartem** [comment]
What is the best place to create this visitor? 
Create new visitors/select_expr_aliases.rs near https://github.com/apache/datafusion-sqlparser-rs/blob/main/src/ast/visitor.rs? Or somewhere in datafustion repo?

**Comment by alamb** [comment]
> What is the best place to create this visitor?

I am not 100% sure about an alias

In general I think code that would create unique aliases would belong in the planner somewhere (`SqlToRel`): https://github.com/apache/datafusion/blob/main/datafusion/sql/src/planner.rs

**Comment by RafaelHerrero** [comment]
I ran into this while running TPC-DS benchmarks against a SQL engine we're building on DataFusion. Opened a PR that adds auto-aliasing in the SQL planner: #21126

**Comment by alamb** [comment]
We had to revert this -- see https://github.com/apache/datafusion/pull/21126#issuecomment-4155559723

I'll try and make a PR to update the extended tests too
